test/drivers/test_revd2.cc

```diff
@@ -0,0 +1,111 @@
+        bool cond_check = false;
+        int64_t p = 10;
+        int64_t passes_per_iteration = 10;
+
+        // Make subroutine objects
+        // Stabilization Constructor - Choose PLU
+        RandLAPACK::PLUL<T> Stab(cond_check, verbosity);
+        // RowSketcher constructor - Choose default (rs1)
+        RandLAPACK::RS<T> RS(Stab, state, p, passes_per_iteration, verbosity, cond_check);
+        // Orthogonalization Constructor - Choose HouseholderQR
+        RandLAPACK::HQRQ<T> Orth_RF(cond_check, verbosity);
+        // RangeFinder constructor - Choose default (rf1)
+        RandLAPACK::RF<T> RF(RS, Orth_RF, verbosity, cond_check);
+        // REVD2 constructor
+        RandLAPACK::REVD2<T> REVD2(RF, std::numeric_limits<double>::epsilon(), 10, state, verbosity);
+
+        k = k_start;
+        // Regular QB2 call
+        REVD2.call(m, A, k, V, e);
+
+        std::vector<T> E(k * k, 0.0);
+        std::vector<T> Buf (m * k, 0.0);
+
+        T* V_dat = V.data();
+        T* E_dat = E.data();
+        T* Buf_dat = Buf.data();
+
+        // Construnct A_hat = U1 * S1 * VT1
+
+        // Turn vector into diagonal matrix
+        RandLAPACK::util::diag(k, k, e, k, E);
+        // V * E = Buf
+        blas::gemm(Layout::ColMajor, Op::NoTrans, Op::NoTrans, m, k, k, 1.0, V_dat, m, E_dat, k, 0.0, Buf_dat, m);
+        // A - Buf * V' - should be close to 0
+        blas::gemm(Layout::ColMajor, Op::NoTrans, Op::Trans, m, m, k, 1.0, Buf_dat, m, V_dat, m, -1.0, A_cpy_dat, m);
+
+        T norm_0 = lapack::lange(Norm::Fro, m, m, A_cpy_dat, m);
+        printf("||A - VEV'||_F / ||A||_F:  %e\n", norm_0 / norm_A);
+        ASSERT_NEAR(norm_0 / norm_A, err_expectation, 10 * err_expectation);
+        ASSERT_NEAR(k, rank_expectation, std::numeric_limits<double>::epsilon());
+
+        printf("|===================================TEST REVD2 GENERAL END===================================|\n");
+    }
+};
+
+TEST_F(TestREVD2, SimpleTest)
+{ 
+    // Generate a random state
+    auto state = RandBLAS::base::RNGState(0, 0);
+    // Rank estimation must be 80 - underestimation - starting with very small rank
+    test_REVD2_general<double>(1000, 100, 1, std::make_tuple(0, std::pow(10, 8), false), state, 64, std::pow(10, -13));
+    // Rank estimation must be 80 - underestimation
+    test_REVD2_general<double>(1000, 100, 10, std::make_tuple(0, std::pow(10, 8), false), state, 80, std::pow(10, -13));
+    // Rank estimation must be 60 - overestimation
+    test_REVD2_general<double>(1000, 100, 10, std::make_tuple(0, std::pow(10, 2), false), state, 160, std::pow(10, -13));
+    // Rank estimation must be 160 - slight overestimation
+    test_REVD2_general<double>(1000, 159, 10, std::make_tuple(0, std::pow(10, 2), false), state, 160, std::pow(10, -13));
+    // Numerically rank deficient matrix - expecting rank estimate = m
+    test_REVD2_general<double>(1000, 1000, 10, std::make_tuple(0, std::pow(10, 2), false), state, 1000, std::pow(10, -13));
```

Comment:
This last test is pretty expensive. Maybe reduce the size to 100-by-100?

Also: don't use a single ``TEST_F`` fixture, use several. If a test fails then we want the gtest log to tell us where we need to look in the code. Right now if one of these tests fail then gtest will be like "well something's broke but good luck my man."

Reply:
Fixed.